from_json/date_test.ts

```diff
@@ -1,72 +1,81 @@
 // Copyright 2018-2020 Gamebridge.ai authors. All rights reserved. MIT license.
 
 import { assert, assertEquals, test } from "../test_deps.ts";
-import { createDateStrategy, ISODateFromJson } from "./date.ts";
+import { createDateStrategy, iso8601Date } from "./date.ts";
 import { Serializable, SerializeProperty } from "../mod.ts";
 
 test({
   name: "createDateStrategy creates strategy from regex",
   fn() {
     const dateStrategy = createDateStrategy(/^(\d{4})-(\d{2})-(\d{2})$/);
-    const testJson = `{"date":"2099-11-25","not_a_date":"Hello world"}`;
+    const testJson = `{"date":"2099-11-25"}`;
     const testObj = JSON.parse(testJson, (_, v) => dateStrategy(v));
     assert(testObj.date instanceof Date);
     assertEquals(testObj.date.getFullYear(), 2099);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-05-24T15:54:14.876Z",
+  name: "Will not serialize non date strings",
   fn() {
-    const testJson =
-      `{"date":"2020-05-24T15:54:14.876Z","not_a_date":"Hello world"}`;
-    const testObj = JSON.parse(testJson, (_, v) => ISODateFromJson(v));
+    class Test extends Serializable {
+      @SerializeProperty({
+        fromJsonStrategy: iso8601Date,
+      })
+      date!: Date | string;
+    }
+    const testObj = new Test().fromJson(`{"date":"I am not a date!"}`);
+    assertEquals(typeof testObj.date, "string");
+    assertEquals(testObj.date, "I am not a date!");
+  },
+});
+
+test({
+  name: "iso8601Date parses with milliseconds - 2020-12-31T12:00:00.300Z",
+  fn() {
+    const testJson = `{"date":"2020-12-31T12:00:00.300Z"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
     assert(testObj.date instanceof Date);
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.300Z");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-12-31T23:00:00+01:00",
+  name: "iso8601Date parses without milliseconds - 2020-12-31T12:00:00Z",
   fn() {
-    const testJson =
-      `{"date":"2020-12-31T23:00:00+01:00","not_a_date":"Hello world"}`;
-    const testObj = JSON.parse(testJson, (_, v) => ISODateFromJson(v));
+    const testJson = `{"date":"2020-12-31T12:00:00"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
     assert(testObj.date instanceof Date);
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-12-31T23:00:00",
+  name: "iso8601Date parses with added timezone - 2020-12-31T12:00:00Z-6",
```

Comment:
I don't think single-digit offsets are technically part of the ISO standard, I believe that a proper year offset needs to be left-padded with 0s